director/stat.go

```diff
@@ -0,0 +1,245 @@
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package director
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/pelicanplatform/pelican/common"
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/utils"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	objectMetadata struct {
+		ServerAd      common.ServerAd `json:"server_ad"`
+		Checksum      string          `json:"checksum"`
+		ContentLength int             `json:"content_length"`
+	}
+
+	timeoutError struct {
+		Message string
+	}
+
+	notFoundError struct {
+		Message string
+	}
+
+	cancelledError struct {
+		Message string
+	}
+
+	// A struct to implement `object stat`, by querying against origins with namespaces match the prefix of an object name
+	// and return origins that have the object
+	ObjectStat struct {
+		ReqHandler func(objectName string, originAd common.ServerAd, timeout time.Duration, maxCancelCtx context.Context) (*objectMetadata, error)
+		Query      func(objectName string, cancelContext context.Context, mininum, maximum int) ([]*objectMetadata, string, error)
+	}
+)
+
+func (e timeoutError) Error() string {
+	return e.Message
+}
+
+func (e notFoundError) Error() string {
+	return e.Message
+}
+
+func (e cancelledError) Error() string {
+	return e.Message
+}
+
+func (meta objectMetadata) String() string {
+	return fmt.Sprintf("Object Meatadata: From server %q at %q.\nContent-length:%d\nChecksum: %s\n",
+		meta.ServerAd.Name,
+		meta.ServerAd.URL.String(),
+		meta.ContentLength,
+		meta.Checksum,
+	)
+}
+
+// Initialize a new stat instance and set default method implementations
+func NewObjectStat() *ObjectStat {
+	stat := &ObjectStat{}
+	stat.ReqHandler = stat.sendHeadReqToOrigin
+	stat.Query = stat.queryOriginsForObject
+	return stat
+}
+
+// Implementation of sending a HEAD request to an origin for an object
+func (stat *ObjectStat) sendHeadReqToOrigin(objectName string, originAd common.ServerAd, timeout time.Duration, maxCancelCtx context.Context) (*objectMetadata, error) {
+	tokenConf := utils.TokenConfig{
+		Lifetime:     time.Minute,
+		TokenProfile: utils.WLCG,
+		Audience:     []string{originAd.URL.String()},
+		Subject:      originAd.URL.String(),
+		// Federation as the issuer
+		Issuer: param.Server_ExternalWebUrl.GetString(),
+	}
+	tokenConf.AddRawScope("storage.read:/")
```

Comment:
If we take the token from the user, wouldn't that token only valid to a whichever origin that issues the token? If that's the case, then I think we lose the value to `stat` an object as a user can simply curl that origin and get a response.

Reply:
I think the missing piece is many of the origins accept/prefer issuers beyond their local one.  LIGO, for example, has a centralized issuer.  I would expect a part of the namespace that has multiple origin services to have a common issuer for all origins that is known to the issuer.

Centralized issuers are heavily used in topology currently.  They are not particularly configurable via Pelican currently - but I assume this will get easier going forward.

Reply:
That makes sense to me. I guess we could circle back to add token based auth once we start using centralized issuer in Pelican